### PR TITLE
SAAS-3405: Added Lex Audio Logs Encrypted Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Audio Logs Encrypted',
+    category: 'Lex',
+    domain: 'Content Delivery',
+    description: 'Ensure that Amazon Lex audio logs are encrypted',
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:8]: Ensure that Amazon Lex audio logs are encrypted using desired KMS encryption level.

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    more_info: 'For audio logs you use default encryption on your S3 bucket or specify an AWS KMS key to encrypt your audio objects. Even if your S3 bucket uses default encryption you can still specify a different AWS KMS key to encrypt your audio objects.',
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:9]:
```suggestion
    more_info: 'For audio logs you use default encryption on your S3 bucket or specify an AWS KMS key to encrypt your audio objects. Even if your S3 bucket uses default encryption you can still specify a different AWS KMS key to encrypt your audio objects for enhanced security.',
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    link: 'https://docs.aws.amazon.com/lex/latest/dg/conversation-logs-encrypting.html',
+    recommended_action: 'Create Lex audio logs with customer-manager keys (CMKs) present in your account',
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:11]:
```suggestion
    recommended_action: 'Encrypt Lex audio logs with customer-manager keys (CMKs) present in your account',
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    apis: ['LexModelsV2:listBots', 'LexModelsV2:listBotAliases', 'LexModelsV2:describeBotAlias',
+        'KMS:describeKey', 'KMS:listKeys', 'STS:getCallerIdentity'],
+    settings: {
+        audio_logs_desired_encryption_level: {
+            name: 'Lex Audio Logs Desired Encryption Level',
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:16]:
```suggestion
            name: 'Lex Audio Logs Target Encryption Level',
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            description: 'In order (lowest to highest) sse=S3-SSE; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
+            regex: '^(sse|awscmk|externalcmk|cloudhsm)$',
+            default: 'awscmk'
+        }
+    },
+
+    run: function(cache, settings, callback) {
+
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:24]:
```suggestion
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        var awsOrGov = helpers.defaultPartition(settings);
+        var accountId = helpers.addSource(cache, source, ['sts', 'getCallerIdentity', defaultRegion, 'data']);
+
+        var config = {
+            desiredEncryptionLevelString: settings.audio_logs_desired_encryption_level || this.settings.audio_logs_desired_encryption_level.default
+        };
+
+        var desiredEncryptionLevel = helpers.ENCRYPTION_LEVELS.indexOf(config.desiredEncryptionLevelString);
+        var currentEncryptionLevel;
+
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:39]: Add a loop for all lexv2models regions

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        var listBots = helpers.addSource(cache, source,
+            ['lexmodelsv2', 'listBots', region]);
+            
+        if (!listBots) return callback(null, results, source);
+
+        if (listBots.err || !listBots.data) {
+            helpers.addResult(results, 3,
+                'Unable to query for  Lex bots: ' + helpers.addError(listBots));
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:47]: Use string interpolation

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        var listKeys = helpers.addSource(cache, source,
+            ['kms', 'listKeys', region]);
+
+        if (!listKeys || listKeys.err || !listKeys.data) {
+            helpers.addResult(results, 3,
+                `Unable to list KMS keys: ${helpers.addError(listKeys)}`, region);
+            return callback(null, results, source);
+        }
+
+        async.each(listBots.data, function(bot, cb){
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:65]: Convert this to for loop

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            if (!bot.botId) return cb();
+
+            var listBotAliases = helpers.addSource(cache, source,
+                ['lexmodelsv2', 'listBotAliases', region, bot.botId]);
+                // console.log(listBotAliases.data);
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:70]:
```suggestion
                ['lexmodelsv2', 'listBotAliases', region, bot.botId]);
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+            if (!listBotAliases || listBotAliases.err || !listBotAliases.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:72]:
```suggestion
            if (!listBotAliases || listBotAliases.err || !listBotAliases.data || !listBotAliases.data.botAliasSummaries) {
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                helpers.addResult(results, 3,
+                    'Unable to query for Lex bot aliases: ' + bot.botId + ': ' + helpers.addError(listBotAliases), region);
+                return cb();
+            }
+
+            if (!listBotAliases.data.botAliasSummaries || !listBotAliases.data.botAliasSummaries.length) {
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:78]:
```suggestion
            if (!listBotAliases.data.botAliasSummaries.length) {
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                helpers.addResult(results, 3,
+                    'Unable to query for Lex bot aliases descriptions: '  + helpers.addError(listBotAliases), region);
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:80]: Add PASS result with no bot aliases found

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                return cb();
+            }
+
+            for (let alias of listBotAliases.data.botAliasSummaries) {
+
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:85]:
```suggestion
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+                return cb();
+            }
+
+            for (let alias of listBotAliases.data.botAliasSummaries) {
+           
+                var resource = `arn:${awsOrGov}:lex:${region}:${accountId}:bot/${alias.botAliasId}`;
+
+                var describeBotAlias = helpers.addSource(cache, source,
+                    ['lexmodelsv2', 'describeBotAlias', region, alias.botAliasId]);
+                    // console.log(JSON.stringify(describeBotAlias , null, 2));
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:90]:
```suggestion
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+                if (!describeBotAlias ||
+                    describeBotAlias.err ||
+                    !describeBotAlias.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to get Lex bot aliases description: ' + alias.botAliasId + ': ' + helpers.addError(describeBotAlias), region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:96]:
```suggestion
                        'Unable to query Lex bot alias: ' + helpers.addError(describeBotAlias), region, resource);
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+                    continue;
+                }
+
+                if (!describeBotAlias.data.conversationLogSettings) {
+                    helpers.addResult(results, 0,
+                        'Unable to get Lex conversation log settings: ' + alias.botAliasId + ': ' + helpers.addError(describeBotAlias), region, resource);
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:102]: Bot alias does not have any audio logs configured

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+                    continue;
+                }
+
+                if (describeBotAlias.data.conversationLogSettings  &&
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:106]: Add login to loop describeBotAlias.data.conversationLogSettings.audioLogSettings

File: plugins/aws/lex/lexAudioLogsEncrypted.spec.js [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+                "enabled": true
+            }
+        },
+        "conversationLogSettings": {
+            "textLogSettings": [
+                {
+                    "enabled": true,
+                    "destination": {
+                        "cloudWatch": {
+                            "cloudWatchLogGroupArn": "arn:aws:logs:us-east-1:101363889637:log-group:mine1",
```
Comment (AkhtarAmir) [inline plugins/aws/lex/lexAudioLogsEncrypted.spec.js:52]:
```suggestion
                            "cloudWatchLogGroupArn": "arn:aws:logs:us-east-1:000011112222:log-group:mine1",
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 19]
```diff
@@ -0,0 +1,156 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Audio Logs Encrypted',
+    category: 'Lex',
+    domain: 'Content Delivery',
+    description: 'Ensure that Amazon Lex audio logs are encrypted using desired KMS encryption leve',
```
Comment (sadeed12345) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:8]:
```suggestion
    description: 'Ensure that Amazon Lex audio logs are encrypted using desired KMS encryption level',
```

File: collectors/aws/collector.js [hunk 20]
```diff
@@ -1902,7 +1923,7 @@ var collect = function(AWSConfig, settings, callback) {
                     LocalAWSConfig.region = region;
 
                     if (callObj.override) {
-                        collectors[serviceLower][callKey](LocalAWSConfig, collection, function() {
+                        collectors[serviceLower][callKey](LocalAWSConfig, collection, retries, function() {
```
Comment (dipsubha06) [inline collectors/aws/collector.js:1926]:
```suggestion
                        collectors[serviceLower][callKey](LocalAWSConfig, collection, function() {
```

File: collectors/aws/collector.js [hunk 21]
```diff
@@ -2065,7 +2086,7 @@ var collect = function(AWSConfig, settings, callback) {
                             if (callObj.signatureVersion) LocalAWSConfig.signatureVersion = callObj.signatureVersion;
 
                             if (callObj.override) {
-                                collectors[serviceLower][callKey](LocalAWSConfig, collection, function() {
+                                collectors[serviceLower][callKey](LocalAWSConfig, collection, retries, function() {
```
Comment (dipsubha06) [inline collectors/aws/collector.js:2089]:
```suggestion
                                collectors[serviceLower][callKey](LocalAWSConfig, collection, function() {
```

File: plugins/aws/lex/lexAudioLogsEncrypted.js [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+            if (!listBots) return rcb();
+
+            if (listBots.err || !listBots.data) {
+                helpers.addResult(results, 3,
+                    'Unable to query for  Lex bots: ' + helpers.addError(listBots),region);
+                return rcb();
+            }
+
+            if (!listBots.data.length) {
+                helpers.addResult(results, 0, 'No  Lex bots found'),region;
```
Comment (dipsubha06) [inline plugins/aws/lex/lexAudioLogsEncrypted.js:52]:
```suggestion
                helpers.addResult(results, 0, 'No  Lex bots found',region);
```

**Review by dipsubha06 — APPROVED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**